pkg/sources/forward/data_forward.go

```diff
@@ -258,49 +259,57 @@ func (df *DataForward) forwardAChunk(ctx context.Context) error {
 		return nil
 	}
 
-	// reset the idle handler because we have read messages
-	df.srcIdleHandler.Reset()
-	metrics.ReadDataMessagesCount.With(map[string]string{
-		metrics.LabelVertex:             df.vertexName,
-		metrics.LabelPipeline:           df.pipelineName,
-		metrics.LabelVertexType:         string(dfv1.VertexTypeSource),
-		metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
-		metrics.LabelPartitionName:      df.reader.GetName()},
-	).Add(float64(len(readMessages)))
-
-	metrics.ReadMessagesCount.With(map[string]string{
+	// Only expose read processing time metric when there are messages and no error
+	metrics.ReadProcessingTime.With(map[string]string{
 		metrics.LabelVertex:             df.vertexName,
 		metrics.LabelPipeline:           df.pipelineName,
 		metrics.LabelVertexType:         string(dfv1.VertexTypeSource),
 		metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
 		metrics.LabelPartitionName:      df.reader.GetName(),
-	}).Add(float64(len(readMessages)))
+	}).Observe(float64(time.Since(readStart).Microseconds()))
+	// reset the idle handler because we have read messages
+	df.srcIdleHandler.Reset()
 
 	// store the offsets of the messages we read from source
-
-	// store the offsets of the messages we read from ISB
 	var readOffsets = make([]isb.Offset, len(readMessages))
 	for idx, m := range readMessages {
-		totalBytes += len(m.Payload)
+		totalBytes := len(m.Payload)
+
+		metrics.ReadDataMessagesCount.With(map[string]string{
+			metrics.LabelVertex:             df.vertexName,
+			metrics.LabelPipeline:           df.pipelineName,
+			metrics.LabelVertexType:         string(dfv1.VertexTypeSource),
+			metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)),
+			metrics.LabelPartitionName:      strconv.Itoa(int(m.ReadOffset.PartitionIdx())),
```

Comment:
This is the first time we use partition index as partition name? and I think it's only for source. Can we introduce a new label say metrics.LabelSourcePartitionIndex, for better naming?

Reply:
We were using partition idx as partition name in kafka source. Since this is only for source, thought of keeping it as partition name only. Because I think for sink, having `metrics.LabelSourcePartitionIndex` would then be redundant as `LabelPartitionName`(below) would already capture the partition idx of buffer. 
`metrics.LabelPartitionName:      df.fromBufferPartition.GetName(),`

```
	for i := 0; i < av.GetPartitionCount(); i++ {
		r = append(r, GenerateBufferName(namespace, pipeline, av.Name, i))
	}
```

Reply:
Got your point. I'm inclined to go with option b. (metrics.LabelPartitionName: "source-partition-" + strconv.Itoa(int(m.ReadOffset.PartitionIdx()))